From 3363252994d7e84cc0152c4b57c43dd9f79cd5f2 Mon Sep 17 00:00:00 2001
From: Yurii Rashkovskii <me@yrashk.com>
Date: Thu, 22 Feb 2018 13:45:32 +0700
Subject: [PATCH] Problem: UI shows "No matches" while requesting issues

This is quite confusing, as there *are* matches, they just weren't
fetched yet.

Solution: display a rotating spinner while they are loading
---
 .../src/dashboard-page/dashboard-page-style.html   |  5 ++
 .../webapp/src/dashboard-page/dashboard-page.html  | 19 ++++--
 sit-web/webapp/src/spinner.html                    | 74 ++++++++++++++++++++++
 3 files changed, 94 insertions(+), 4 deletions(-)
 create mode 100644 sit-web/webapp/src/spinner.html

diff --git a/sit-web/webapp/src/dashboard-page/dashboard-page-style.html b/sit-web/webapp/src/dashboard-page/dashboard-page-style.html
index 2561561..8ae1ca5 100644
--- a/sit-web/webapp/src/dashboard-page/dashboard-page-style.html
+++ b/sit-web/webapp/src/dashboard-page/dashboard-page-style.html
@@ -10,6 +10,11 @@
                 background-color: var(--sit-header-color);
                 color: #fff;
             }
+
+            sit-spinner {
+                margin: 0 auto;
+                width: 50px;
+            }
         </style>
     </template>
 </dom-module>
diff --git a/sit-web/webapp/src/dashboard-page/dashboard-page.html b/sit-web/webapp/src/dashboard-page/dashboard-page.html
index 9bf1dc6..0e40a3e 100644
--- a/sit-web/webapp/src/dashboard-page/dashboard-page.html
+++ b/sit-web/webapp/src/dashboard-page/dashboard-page.html
@@ -8,7 +8,7 @@
 <link rel="import" href="/bower_components/neon-animation/web-animations.html">
 <link rel="import" href="/bower_components/iron-flex-layout/iron-flex-layout-classes.html">
 
-
+<link rel="import" href="../spinner.html">
 <link rel="import" href="dashboard-page-style.html">
 <link rel="import" href="issue.html">
 
@@ -44,14 +44,20 @@
         <iron-ajax auto url="/api/issues/{{ encodedFilter }}/{{ query }}" handle-as="json"
                    last-response="{{ issues }}"
                    on-error="handleError"
+                   on-request="handleRequest"
                    on-response="handleResponse"></iron-ajax>
        <template is="dom-if" if="[[ badQuery ]]">
            <h1>Wrong query, check the syntax (<a href="http://jmespath.org/specification.html" target="_blank">JMESPath specification</a>).</h1>
        </template>
-       <template is="dom-if" if="[[ emptyResult ]]">
-           <h1>No matches</h1>
+       <template is="dom-if" if="[[ requesting ]]">
+           <sit-spinner></sit-spinner>
+       </template>
+       <template is="dom-if" if="[[ !requesting ]]">
+           <template is="dom-if" if="[[ emptyResult ]]">
+               <h1>No matches</h1>
+           </template>
        </template>
-       <div id="results" class="layout vertical">
+      <div id="results" class="layout vertical">
         <template is="dom-repeat" items="{{ issues }}">
             <sit-issue issue="{{item}}"></sit-issue>
         </template>
@@ -141,8 +147,13 @@
                 this.query = "{id: id, summary: summary, state: state, comments: (comments || `[]`), merge_request: merge_request || false, last_updated_timestamp: last_updated_timestamp}";
             }
 
+            handleRequest() {
+                this.set("requesting", true);
+            }
+
             handleResponse(evt, req) {
                 this.set("badQuery", false);
+                this.set("requesting", false);
             }
 
             handleError(evt, req) {
diff --git a/sit-web/webapp/src/spinner.html b/sit-web/webapp/src/spinner.html
new file mode 100644
index 0000000..ea4787a
--- /dev/null
+++ b/sit-web/webapp/src/spinner.html
@@ -0,0 +1,74 @@
+<link rel="import" href="/bower_components/polymer/polymer-element.html">
+<dom-module id="sit-spinner">
+    <template>
+        <style>
+            :host {
+                display: block;
+            }
+            @-webkit-keyframes rotating /* Safari and Chrome */ {
+                from {
+                    -webkit-transform: rotate(0deg);
+                    -o-transform: rotate(0deg);
+                    transform: rotate(0deg);
+                    transform-origin: center;
+                }
+                to {
+                    -webkit-transform: rotate(360deg);
+                    -o-transform: rotate(360deg);
+                    transform: rotate(360deg);
+                    transform-origin: center;
+                }
+            }
+            @keyframes rotating {
+                from {
+                    -ms-transform: rotate(0deg);
+                    -moz-transform: rotate(0deg);
+                    -webkit-transform: rotate(0deg);
+                    -o-transform: rotate(0deg);
+                    transform: rotate(0deg);
+                    transform-origin: center;
+
+                }
+                to {
+                    -ms-transform: rotate(360deg);
+                    -moz-transform: rotate(360deg);
+                    -webkit-transform: rotate(360deg);
+                    -o-transform: rotate(360deg);
+                    transform: rotate(360deg);
+                    transform-origin: center;
+                }
+            }
+            .spin {
+                -webkit-animation: rotating 14s linear infinite;
+                -moz-animation: rotating 1s linear infinite;
+                -ms-animation: rotating 1s linear infinite;
+                -o-animation: rotating 1s linear infinite;
+                animation: rotating 1s linear infinite;
+            }
+        </style>
+        <svg class="spin" width="50" height="50" viewBox="0 0 177.61257 177.61256" >
+            <g id="layer1"
+               transform="translate(-118.2661,-79.359196)">
+                <g id="g1664"
+                   transform="matrix(5.2934299,0,0,5.2934299,926.765,633.24869)">
+                    <g transform="matrix(0.35277777,0,0,-0.35277777,-126.01797,-82.889473)"
+                       id="g36">
+                        <path
+                                id="path38"
+                                style="fill:#43b3c2;fill-opacity:1;fill-rule:nonzero;stroke:none"
+                                d="M 0,0 V 28.182 L 14.091,14.091 Z m -28.182,-1.761 c -8.754,0 -15.851,7.097 -15.851,15.852 0,8.755 7.097,15.852 15.851,15.852 8.755,0 15.853,-7.097 15.853,-15.852 0,-8.755 -7.098,-15.852 -15.853,-15.852 m 0,-26.42 -14.089,14.09 h 28.18 z m 0,84.544 14.091,-14.091 h -28.18 z M -56.362,0 -70.453,14.091 -56.362,28.182 Z m 49.317,35.227 v 26.42 h -42.272 v -26.42 h -26.42 V -7.045 h 26.42 v -26.42 h 42.272 v 26.42 h 26.42 v 42.272 z" >
+                        </path>
+                    </g>
+                </g>
+            </g>
+        </svg>
+    </template>
+    <script>
+        class Spinner extends Polymer.Element {
+            static get is() {
+                return 'sit-spinner';
+            }
+        }
+        customElements.define(Spinner.is, Spinner)
+    </script>
+</dom-module>
\ No newline at end of file
-- 
2.15.0

